### ISSUE TITLE
When verbatim output crosses page boundary some characters have active catcode which may cause problems in headers if `\scantokens` is used

### ISSUE THREAD

**Issue by jfbu** [latex3/latex2e#600]
## Brief outline of the bug

As the title says. In the example below I show the effect on a macro using `\xspace`: the comma in the header is not recognized as such by `xspace` and an extra space appears.

![Capture d’écran 2021-06-25 à 20 51 16](https://user-images.githubusercontent.com/2589111/123472178-20de2580-d5f7-11eb-8275-7412d754e434.png)

(edit: the comma in the screen shot is of catcode 12, but `\xspace`  does a `\scantokens` and then it will be interpreted as being of catcode 13; I am sorry about the imprecise wording in setting up the MWE; in my old file not MWE i had noted the issue was a `\scantokens` done by `\xspace`)

I have encountered this some years ago and fixed it by adding ``\catcode`,12\relax`` in my use case to my running header but did not take the time yet to produce a minimal example (it does take time)  although I knew the issue was caused by an active comma.

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}
\documentclass{article}

\usepackage{xspace}

\newcommand\foo{possible unwanted space before comma\xspace}
\protected\def\checkcomma{(its catcode here is \the\catcode`\,\relax)}

\pagestyle{headings}

\begin{document}
\markboth{\foo, \checkcomma}{\foo, \checkcomma}
\begin{verbatim}
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
\end{verbatim}
\end{document}

```

## Log file (required) and possibly PDF file  

[testcommainheader.log](https://github.com/latex3/latex2e/files/6718214/testcommainheader.log)

**Comment by jfbu** [comment]
Update : the comma is of catcode 12 but `\xspace`  does scantokens. Forgot to mention this from my old notes. I will change my title...

**Comment by blefloch** [comment]
I suspect `\xspace` should have two exception lists, one for characters tested with `\if\expandafter\noexpand\csname @let@token\endcsname\noexpand#1` and one for control sequences or active characters, tested with `\expandafter\ifx\csname @let@token\endcsname#1`.

But perhaps something needs to be changed concerning the catcode regime with which we process headers.

**Comment by car222222** [comment]
The catcode regime certainly needs to be "correct" throughout the Output Routine, not just when processing headers, etc. 

But there may also be other processing, that can happen outside the Output Routine and also needs attention since it can similarly be adversely affected by happening during non-standard catcode regimes.

**Comment by jfbu** [comment]
Indeed, and besides `\scantokens` as focused upon in my initial report due to `\xspace` internals, there is also the case of `\input`.  Here is another mwe:

```latex
\documentclass{article}
\begin{filecontents*}{\jobname-1}
\textbf{Preface}
\end{filecontents*}
\begin{filecontents*}{\jobname-2}
\textbf{Conclusion}
\end{filecontents*}
\makeatletter
\DeclareRobustCommand{\InputHeaderFile}{\input{\jobname-\number\c@page}}
\makeatother

\pagestyle{headings}
\markboth{\InputHeaderFile}{\InputHeaderFile}

\begin{document}
\begin{verbatim}
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
x
\end{verbatim}
\end{document}
```

which produces 
```
“textbf–Preface ̋
```
at top of page 1. And there are other circumstances than `verbatim` where the catcode regime may be altered at time of pagebreak.

**Comment by pietvo** [comment]
You get the same kind of problem when you put `\input{somefile}` in a `shipout` hook (e.g `shipout/foreground`). If the `verbatim` is active at the pagebreak, then the text of `somefile` will be processed in `verbatim` mode, instead of being interpreted as LaTeX code.

The output routine should reset the catcodes of characters that are changed in verbatim(-like) environments to their default values, and maybe some additional characters too.

**Comment by FrankMittelbach** [comment]
I agree with you @pietvo, since the OR marks a context change it should normalize the context (and provide a way for special environments to initiate further resetting if necessary. That is actually on the agenda, but should go hand in hand with other adjustments to the OR which is why it is still not done.

**Comment by pietvo** [comment]
> I agree with you @pietvo, since the OR marks a context change it should normalize the context (and provide a way for special environments to initiate further resetting if necessary. ...

Yes, there should be hooks at entry/exit of the OR.
